sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java

```diff
@@ -0,0 +1,535 @@
+    private String current;
+
+    public BoundedMongoDbReader(String uri, String database, String collection, String filter) {
+      this.uri = uri;
+      this.database = database;
+      this.collection = collection;
+      this.filter = filter;
+    }
+
+    @Override
+    public boolean start() {
+      client = new MongoClient(new MongoClientURI(uri));
+
+      MongoDatabase mongoDatabase = client.getDatabase(database);
+
+      MongoCollection<Document> mongoCollection = mongoDatabase.getCollection(collection);
+
+      if (filter == null) {
+        cursor = mongoCollection.find().iterator();
+      } else {
+        Document bson = Document.parse(filter);
+        cursor = mongoCollection.find(bson).iterator();
+      }
+
+      return advance();
+    }
+
+    @Override
+    public boolean advance() {
+
+      if (cursor.hasNext()) {
+        current = cursor.next().toJson();
+        return true;
+      } else {
+        return false;
+      }
+    }
+
+    @Override
+    public BoundedSource getCurrentSource() {
+      return null;
+    }
+
+    @Override
+    public Object getCurrent() {
+      return current;
+    }
+
+    @Override
+    public void close() {
+      cursor.close();
+      client.close();
+    }
+
+  }
+
+  /**
+   * A {@link PTransform} to write to a MongoDB database.
+   */
+  public static class Write extends PTransform<PCollection<String>, PDone> {
+
+    public Write withUri(String uri) {
+      return new Write(uri, database, collection);
+    }
+
+    public Write withDatabase(String database) {
+      return new Write(uri, database, collection);
+    }
+
+    public Write withCollection(String collection) {
+      return new Write(uri, database, collection);
+    }
+
+    protected String uri;
+    protected String database;
+    protected String collection;
+
+    private Write() {}
+
+    private Write(String uri, String database, String collection) {
+      this.uri = uri;
+      this.database = database;
+      this.collection = collection;
+    }
+
+    @Override
+    public PDone apply(PCollection<String> input) {
+      input.apply(ParDo.of(new MongoDbWriter(uri, database, collection)));
+      return PDone.in(input.getPipeline());
+    }
+
+    @Override
+    public void validate(PCollection<String> input) {
+      Preconditions.checkNotNull(uri, "uri");
+      Preconditions.checkNotNull(database, "database");
+      Preconditions.checkNotNull(collection, "collection");
+    }
+
+    private static class MongoDbWriter extends DoFn<String, Void> {
+
+      private String uri;
+      private String database;
+      private String collection;
+
+      private MongoClient client;
+
+      public MongoDbWriter(String uri, String database, String collection) {
+        this.uri = uri;
+        this.database = database;
+        this.collection = collection;
+      }
+
+      @StartBundle
+      public void startBundle(Context c) throws Exception {
+        if (client == null) {
```

Comment:
Please use the @Setup and @Teardown methods to create and destroy the client. That's more efficient, because they're invoked just once per DoFn instance, and then you don't need null checks (it will *definitely* be null in @Setup, and *definitely* non-null in @Teardown).